RestHighLevelClient fails to parse ShardSearchFailure on Scroll

When Scrolling, `ShardSearchFailures` returned by ElasticSearch return no `index`, this in turn makes `RestHighLevelClient` fail to parse them, because `Index` [requires name not to be null](https://github.com/elastic/elasticsearch/blob/5.6/core/src/main/java/org/elasticsearch/index/Index.java#L53)

We're seeing the following exceptions:
```
java.io.IOException: Unable to parse response body for Response{requestLine=GET /_search/scroll HTTP/1.1, host=http://unrouted.ds-apicore-misc-01.ds:80, response=HTTP/1.1 200 OK}
	at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:394)
	at org.elasticsearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:361)
	at org.elasticsearch.client.RestHighLevelClient.searchScroll(RestHighLevelClient.java:321)
	...
Caused by: java.lang.NullPointerException
	at java.util.Objects.requireNonNull(Objects.java:203)
	at org.elasticsearch.index.Index.<init>(Index.java:53)
	at org.elasticsearch.action.search.ShardSearchFailure.fromXContent(ShardSearchFailure.java:213)
	at org.elasticsearch.action.search.SearchResponse.fromXContent(SearchResponse.java:297)
	at org.elasticsearch.client.RestHighLevelClient.parseEntity(RestHighLevelClient.java:505)
	at org.elasticsearch.client.RestHighLevelClient.lambda$performRequestAndParseEntity$2(RestHighLevelClient.java:361)
	at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:392)
	... 39 more
```

We're in the process of migrating our `TransportClients` to `RestHighLevelClient`, when using `TransportClient` this exceptions did not occur.

On `TransportClient` this translates to having a response with the following shard failures:
```
shard [_na], reason [RemoteTransportException[[i-0ad52147150c19404][10.64.92.24:9300][indices:data/read/search[phase/query/scroll]]]; nested: SearchContextMissingException[No search context found for id [5862]]; ], cause [SearchContextMissingException[No search context found for id [5862]]
```

**Elasticsearch version** 5.6.3

Edit:

This apparently seems to happen because on Scrolls the `shardTarget` of a `ShardSearchFailure` can be [null](https://github.com/elastic/elasticsearch/blob/5.6/core/src/main/java/org/elasticsearch/action/search/ShardSearchFailure.java#L170), but when parsing itself with `fromXContent` [it always tries to instantiate one with null values](https://github.com/elastic/elasticsearch/blob/5.6/core/src/main/java/org/elasticsearch/action/search/ShardSearchFailure.java#L214).

--- a/core/src/main/java/org/elasticsearch/action/search/ShardSearchFailure.java
+++ b/core/src/main/java/org/elasticsearch/action/search/ShardSearchFailure.java
@@ -131,7 +131,8 @@ public class ShardSearchFailure implements ShardOperationFailedException {
 
     @Override
     public String toString() {
-        return "shard [" + (shardTarget == null ? "_na" : shardTarget) + "], reason [" + reason + "], cause [" + (cause == null ? "_na" : ExceptionsHelper.stackTrace(cause)) + "]";
+        return "shard [" + (shardTarget == null ? "_na" : shardTarget) + "], reason [" + reason + "], cause [" +
+                (cause == null ? "_na" : ExceptionsHelper.stackTrace(cause)) + "]";
     }
 
     public static ShardSearchFailure readShardSearchFailure(StreamInput in) throws IOException {
@@ -210,9 +211,12 @@ public class ShardSearchFailure implements ShardOperationFailedException {
                 parser.skipChildren();
             }
         }
-        return new ShardSearchFailure(exception,
-                new SearchShardTarget(nodeId,
-                        new ShardId(new Index(indexName, IndexMetaData.INDEX_UUID_NA_VALUE), shardId), null, OriginalIndices.NONE));
+        SearchShardTarget searchShardTarget = null;
+        if (nodeId != null) {
+            searchShardTarget = new SearchShardTarget(nodeId,
+                    new ShardId(new Index(indexName, IndexMetaData.INDEX_UUID_NA_VALUE), shardId), null, OriginalIndices.NONE);
+        }
+        return new ShardSearchFailure(exception, searchShardTarget);
     }
 
     @Override

--- a/core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java
+++ b/core/src/test/java/org/elasticsearch/action/search/SearchResponseTests.java
@@ -175,7 +175,7 @@ public class SearchResponseTests extends ESTestCase {
                 ShardSearchFailure parsedFailure = parsed.getShardFailures()[i];
                 ShardSearchFailure originalFailure = failures[i];
                 assertEquals(originalFailure.index(), parsedFailure.index());
-                assertEquals(originalFailure.shard().getNodeId(), parsedFailure.shard().getNodeId());
+                assertEquals(originalFailure.shard(), parsedFailure.shard());
                 assertEquals(originalFailure.shardId(), parsedFailure.shardId());
                 String originalMsg = originalFailure.getCause().getMessage();
                 assertEquals(parsedFailure.getCause().getMessage(), "Elasticsearch exception [type=parsing_exception, reason=" +

--- a/core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java
+++ b/core/src/test/java/org/elasticsearch/action/search/ShardSearchFailureTests.java
@@ -20,6 +20,7 @@
 package org.elasticsearch.action.search;
 
 import org.elasticsearch.action.OriginalIndices;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.common.ParsingException;
 import org.elasticsearch.common.bytes.BytesReference;
 import org.elasticsearch.common.xcontent.ToXContent;
@@ -40,12 +41,14 @@ public class ShardSearchFailureTests extends ESTestCase {
     public static ShardSearchFailure createTestItem() {
         String randomMessage = randomAlphaOfLengthBetween(3, 20);
         Exception ex = new ParsingException(0, 0, randomMessage , new IllegalArgumentException("some bad argument"));
-        String nodeId = randomAlphaOfLengthBetween(5, 10);
-        String indexName = randomAlphaOfLengthBetween(5, 10);
-        String indexUuid = randomAlphaOfLengthBetween(5, 10);
-        int shardId = randomInt();
-        return new ShardSearchFailure(ex,
-                new SearchShardTarget(nodeId, new ShardId(new Index(indexName, indexUuid), shardId), null, null));
+        SearchShardTarget searchShardTarget = null;
+        if (randomBoolean()) {
+            String nodeId = randomAlphaOfLengthBetween(5, 10);
+            String indexName = randomAlphaOfLengthBetween(5, 10);
+            searchShardTarget = new SearchShardTarget(nodeId,
+                    new ShardId(new Index(indexName, IndexMetaData.INDEX_UUID_NA_VALUE), randomInt()), null, null);
+        }
+        return new ShardSearchFailure(ex, searchShardTarget);
     }
 
     public void testFromXContent() throws IOException {
@@ -80,10 +83,10 @@ public class ShardSearchFailureTests extends ESTestCase {
             assertNull(parser.nextToken());
         }
         assertEquals(response.index(), parsed.index());
-        assertEquals(response.shard().getNodeId(), parsed.shard().getNodeId());
+        assertEquals(response.shard(), parsed.shard());
         assertEquals(response.shardId(), parsed.shardId());
 
-        /**
+        /*
          * we cannot compare the cause, because it will be wrapped in an outer
          * ElasticSearchException best effort: try to check that the original
          * message appears somewhere in the rendered xContent

